lld/ELF/Arch/LoongArch.cpp

```diff
@@ -153,6 +154,23 @@ static bool isJirl(uint32_t insn) {
   return (insn & 0xfc000000) == JIRL;
 }
 
+static void handleUleb128(uint8_t *loc, uint64_t val) {
+  const char *err = nullptr;
+  const uint32_t maxcount = 1 + (config->is64 ? 64 : 32) / 7;
+  uint32_t count;
+  uint64_t orig = decodeULEB128(loc, &count, nullptr, &err);
+  if (err)
+    fatal(getErrorLocation(loc) + "could not decode uleb128 value: " + err);
+  if (count > maxcount)
+    errorOrWarn(getErrorLocation(loc) + "extra space for uleb128");
+  uint64_t mask = config->is64 ? -1 : -1u;
+  if (count < maxcount)
+    mask = (1 << 7 * count) - 1;
+
+  val = (orig + val) & mask;
+  encodeULEB128(val, loc, count);
```

Comment:
see `overwriteULEB128`

Reply:
I used `decodeULEB128` before, so I prefer to using `encodeULEB128` here. Furthermore, LoongArch psABI does not guarantee the relocation order between R_LARCH_{ADD,SUB}_ULEB128 (, although usually relocate ADD before SUB). `overwriteULEB128` will return the rest of `val` not writed. Thus, if it write negative value first, `overwriteULEB128(loc, val) >= 0x80` will be true in most case. And I think diagnostic `ULEB128 value ... exceeds available space` is not suitable on LoongArch.

Reply:
Looks like `R_LARCH_ADD_ULEB128` using ADD instead of SET has an interesting trade-off. The relocation overflow cannot be detected while add/sub can now be reordered (though there may be no point for reordering). Technically, one location can have multiple ADD and SUB.